Add a tutorial step covering icons

freakboy3742 opened the pull request:
Adds a step 10 tutorial, describing icons and splash screens.

Also does a light update of other steps and images to reflect recent changes.

Refs beeware/toga#2527
Refs beeware/briefcase#1725
Refs beeware/briefcase-template#111
Refs beeware/briefcase-macOS-app-template#47
Refs beeware/briefcase-android-gradle-template#88

Fixes #317 
Fixes #333

## PR Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

mhsmith commented on docs/tutorial/tutorial-7.rst:
```diff
@@ -342,7 +347,8 @@ the settings for your app::
     [tool.briefcase.app.helloworld]
     formal_name = "Hello World"
     description = "A Tutorial app"
-    icon = "src/helloworld/resources/helloworld"
```
Removal of `icon` depends on the outcome of the discussion in https://github.com/beeware/briefcase-template/pull/111.

mhsmith commented on docs/tutorial/tutorial-10.rst:
```diff
@@ -4,17 +4,479 @@
+So - in addition to configuring Briefcase to use the new icon for the binary and
+installer, we also need to provide updated icons for our app to use at runtime.
+
+If you look in the app's ``src/helloworld/resources`` folder, you can see the
+icons that are being used by Toga at runtime. We can overwrite these files with
+versions from our ``icons`` folder, then update and re-run our app. We use the
```
Having different icons for the app executable and the running app process seems too advanced a topic for a tutorial. And since very few apps will actually want them to be different, leaving the user with two identical icon directories to maintain is giving them extra work for no benefit.

So I suggest keeping the tutorial focused on a single icon set in the default location, and explaining this binary/runtime distinction in the Briefcase icon documentation instead.

freakboy3742 replied:
I agree this is a complex topic to be covering in an introductory tutorial; however, on the basis that something that is hard to explain probably isn't a good design, I think this is evidence that the underlying feature of app icons has issues.

This distinction between "runtime" and "app binary" icons really only matters because Toga is *explicitly* setting a runtime icon. You *can* provide an icon in the construction of `toga.App`; but almost all apps (including the tutorial) fall back onto the default behavior of loading `resources/app_name` as the default icon. 

The app template for each platform *also* includes an icon. This is what gets installed when you do an `update --update-resources`. The current `briefcase-template` setup is essentially using `src/appname/resources/appname` as an *input*, so that it can generate an icon in the app bundle *and* have the same icon available at runtime.

However, this poses 2 problems:
1. As the original icons are in the `src/appname` folder, they are packaged as app resources - which means that a  bundled Windows app will include a macOS `.icns` and GTK `.png` file
2. It requires duplication of the icon file so it can be found both as a runtime resource and by the app binary

If the app is also published as a mobile app, the situation gets worse because of the number of icons needed to support Android and iOS.

If you *don't* explicitly specify an icon at runtime, the icon that is used is the default for the app binary - which is what we're specifying with Briefcase. So, if we modify Toga to *not* explicitly set an icon, and fall back to the default icon for the binary, we don't need to explicitly include the icons as runtime resources. 

This then leaves the case of a Toga app that isn't using Briefcase at all (i.e., `python -m myapp`). A Briefcase app running in dev mode is functionally equivalent to this.

So - I'd like to suggest modifying Toga's app icon behavior:
* If an app explicitly specifies an icon (either by setting `app.icon`, or passing in an icon when constructing the `toga.App` instance) that icon is used.
* If there's no explicit icon, and `sys.executable` is `python`, `python3` or `python3.X`, Toga falls back to the current `resources/app_name` behaviour. 
* If there's no explicit icon and `sys.executable` is anything else, it falls back to the binary's default icon.

This would allow us to modify the default app template to remove the `icon` definition *and* the default icon images in `resources`, and it would avoid shipping macOS icons with Windows apps. The tutorial can then stop after the successful `update --update-resources` pass. It also retains basic Toga branding for the `python -m` case (and, by extension, the `briefcase dev` case).

One additional change could be made to improve the ergonomics of the `briefcase dev` case: Modify Briefcase to include a `BRIEFCASE_APP_ICON` environment variable in the runtime environment if the app configuration specifies a value for `icon`. This would only exist in the dev environment, and would precedence over the 2 fallback options.

I'm not 100% sold on this last change. I'm not all that happy about introducing an explicit dependency between Briefcase and Toga, especially when it's purely for dev mode; but I'm not sure if it's easier to explain that "dev mode doesn't have the app icon", or just include that in the list of things that is different between dev mode and run mode. 

Thoughts?

freakboy3742 replied:
To be clear, the sys.executable variant is as much for Toga branding purposes, rather than run/dev differentiation. The default icons for running Python apps aren't consistent across platforms (and in some cases, quite bad), and don't give any indicator that the app is a GUI (much less Toga) app.

To much lesser extent, it's also a backwards compatibility detail; if you've got a Toga script today with a custom icon, the `resources/app_name` icon will be used. Admittedly, this is a less common user case, and I'm not fundamentally opposed to this being a backwards incompatibility - but given that we get a branding boost as well, I figured we might as well keep it.

mhsmith replied:
Where does the "branding boost" come from? Is it an implicit fourth bullet which says "if none of the above succeed, use the Toga logo"? That seems like an awful lot of "if .. and not" conditions that the user would have to read through to work out what's going on.

The current Toga [icon documentation](https://toga.readthedocs.io/en/stable/reference/api/app.html#toga.App) is much simpler:

> The [icon](https://toga.readthedocs.io/en/stable/reference/api/resources/icons.html#c.IconContent) for the app. If not provided, Toga will attempt to load an icon from resources/app_name, where app_name is defined above. If no resource matching this name can be found, a warning will be printed, and the app will fall back to a default icon.

What if we just change this from "a default icon" to "the icon of the app's executable file", and remove "a warning will be printed"? That would preserve the `resources/app_name` behavior for apps using the previous template, and fall back on the executable's icon for apps using the new template, without making the behavior any more complicated than it already is.

freakboy3742 replied:
The implementation I've got on beeware/toga#2527 clarifies (and documents) the options; I've also removed some cases when the warning is printed. It's probably easiest to continue the discussion there where it's tied to specifics. Once those set of changes are finalised, I'll update the tutorial to match.

freakboy3742 commented:
I've updated the tutorial on the assumption that beeware/toga#2527 (or something close to it) will be merged. This removes all references to "runtime" icons. I've also updated some screenshots from earlier in the tutorial that will now show the yak icon. There's likely a couple more tweaks to come as I sort out the Android icon size issue, and modify splash screens to be icon based.
